File: leetcode/GenerateParentheses/generateParenthesis.java
public class Solution {
    public ArrayList<String> generateParenthesis(int n) {
        // Start typing your Java solution below
        // DO NOT write main() function
        return helper(n, n);
    }
    private ArrayList<String> helper(int left, int right) {
        ArrayList<String> res = new ArrayList<String>();
        if (left != 0 && left <= right) {
            for (String s : helper(left - 1, right)) {
                res.add("(" + s);
            }
        }
        if (right != 0 && left < right) {
            if (right == 1) res.add(")");
            else {
                for (String s : helper(left, right - 1)) {
                    res.add(")" + s);
                }
            }
        }
        return res;
    }
}

File: leetcode/GenerateParentheses/generateParenthesis.cpp
class Solution {
public:
    void helper(int left, int right, string str, vector<string> &ans) {
        if (!left && !right) ans.push_back(str);
        if (left) {
            helper(left - 1, right, str + '(', ans);    
        } 
        if (right > left) {
            helper(left, right - 1, str + ')', ans);
        }
    }
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        helper(n, n, "", ans);
        return ans;
    }
};
